### PR TITLE
Proofs for h-initiality and inductiveness of Nat (and the equivalence of the two properties)

### PR DESCRIPTION
**Pull request by WorldSEnder**
Implements [Homotopy-initial algebras in type theory][1] for natural numbers.

There are several questions to be discussed
 - define containers / W-types and generalize? There should be no technical problem in the generalization
 - `SubstProperties` and `CompProperties` should probably get put into the correct place somewhere in `Cubical.Foundations`?
 - Add the module to export list of other modules. It can not be made accessible from `Cubical.Data.Nat`, as there is a circular dependency:
  > ...
  >   Cubical.Foundations.HLevels
  >   Cubical.Data.Nat
  >   Cubical.Data.Nat.Properties
 - naming! Please critique

 [1]: https://arxiv.org/abs/1504.05531

### CODE REVIEW

**Comment by mortberg** [comment]
Btw, how did the complexity of the proofs compare to the HoTT proofs? Does the use of the interval and Path types simplify things?

File: Cubical/Data/Nat/Algebra.agda [hunk 1]
```diff
@@ -0,0 +1,276 @@
+{-# OPTIONS --cubical --no-exact-split --safe #-}
```
[the remark this answers is not in the document]
Comment (mortberg) [reply Cubical/Data/Nat/Algebra.agda]: "saying **what**"

File: Cubical/Data/Nat/Algebra.agda [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+  morph→section x = record { section = morph ; sec-comm-zero = comm-zero ; sec-comm-suc = λ i n → comm-suc n i } where open NatMorphism x
+  section→morph : NatSection ConstFiberM → NatMorphism N M
+  section→morph x = record { morph = section ; comm-zero = sec-comm-zero ; comm-suc = λ n i → sec-comm-suc i n } where open NatSection x
+  Morph≡Section : NatSection ConstFiberM ≡ NatMorphism N M
+  Morph≡Section = isoToPath (iso section→morph morph→section (λ _ → refl) (λ _ → refl))
+
+  isContrMorph : isContr (NatMorphism N M)
+  isContrMorph = subst isContr Morph≡Section (inhProp→isContr (ind ConstFiberM) (AlgebraPropositionality.SectionProp.S≡T ind))
+
+module Helper {a b} {A : Set a} (B : A → Set b) where
```
[the remark this answers is not in the document]
Comment (WorldSEnder) [reply Cubical/Data/Nat/Algebra.agda]: I see at least one lemma about `subst` in Cubical.Foundations.Transport. Do you think that be a suitable place for now?

[the remark this answers is not in the document]
Comment (mortberg) [reply Cubical/Data/Nat/Algebra.agda]: Yeah, that sounds good!

**Comment by WorldSEnder** [comment]
Working with intervals and paths was very pleasant for deriving the homotopy between two `NatSection` (`S≡T` in SectionProp). On the other hand, working out the groupoid laws needed (`P-zero` and `P-suc` in `HInit→Ind`) did take some time to get right.

**Comment by WorldSEnder** [comment]
>     * I'm not sure what to do about the circular dependencies... If the code cannot be made accessible from Cubical.Data.Nat then maybe the code should go into Cubical.Data.Nat.Properties?

The problem with that is that `Foundations.HLevels` uses `+-comm` from `Nat.Properties` but I want to use `propPi` from `HLevels` leading to a circular import.

**Comment by mortberg** [comment]
> > ```
> > * I'm not sure what to do about the circular dependencies... If the code cannot be made accessible from Cubical.Data.Nat then maybe the code should go into Cubical.Data.Nat.Properties?
> > ```
> 
> The problem with that is that `Foundations.HLevels` uses `+-comm` from `Nat.Properties` but I want to use `propPi` from `HLevels` leading to a circular import.

I see. I don't think HLevels should use +-comm. The main reason to split Nat into Base and Properties was to avoid circular imports in HLevels. It seems like +-comm is only used in one place in HLevels... Can you just inline the proof there and fix the imports so that it only depends on Nat.Base?

**Comment by mortberg** [comment]
I think all of my comments have been fixed now? If so I'll merge

**Comment by WorldSEnder** [comment]
I believe so. Go ahead

**Pull request merged**